### PR TITLE
GH-1467: Support Spring AOT for Native Image

### PR DESCRIPTION
**Pull request by garyrussell**
Resolves https://github.com/spring-projects/spring-amqp/issues/1467

Migrate proxy hints from spring-native.

Also add a proxy hint for `ChannelProxy` when it is also a `PublisherCallbackChannel`.

### CODE REVIEW

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java:43]: Gary, don't we need also these interfaces for each proxy: `SpringProxy.class, Advised.class, DecoratingProxy.class` ?
See my PR in SI: https://github.com/spring-projects/spring-integration/pull/3832/files#diff-b471878878aa8838786d4a93b19f8014d497db6ace582572f834da814720e01bR150.

Also see this new API in SF `6.0`: 
```
runtimeHints.proxies().registerJdkProxy(AopProxyUtils.completeJdkProxyInterfaces(proxyInterfaces))
```

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java
Comment (garyrussell) [reply spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java:43]: I just replicated what we have in spring-native (which works).

https://github.com/spring-projects-experimental/spring-native/blob/2d1fc19c3833e6ce2bd4d342504b180a88407572/spring-native-configuration/src/main/java/org/springframework/amqp/rabbit/annotation/RabbitHints.java#L28-L35

I will find out soon because the compile problem has now been resolved.

Comment (artembilan) [reply spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java:43]: I see. Our logic for this proxy is not about Spring. Agreed:
```
		if (ConfirmType.CORRELATED.equals(this.confirmType) || this.publisherReturns) {
			interfaces = new Class<?>[] { ChannelProxy.class, PublisherCallbackChannel.class };
		}
		else {
			interfaces = new Class<?>[] { ChannelProxy.class };
		}
		return (ChannelProxy) Proxy.newProxyInstance(ChannelProxy.class.getClassLoader(),
				interfaces, new CachedChannelInvocationHandler(connection, targetChannel, channelList,
						transactional));
```
However I I guess we need to register one more proxy for the `ChannelProxy.class` and `PublisherCallbackChannel.class` combination according the logic of that `CachingConnectionFactory` class.
The one for `RabbitListener` annotation is OK. However it might be something like:
```
			RuntimeHintsUtils.registerAnnotation(hints, Scheduled.class);
```
instead ?

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-rabbit/src/main/resources/META-INF/spring/aot.factories
Comment (artembilan) [inline spring-rabbit/src/main/resources/META-INF/spring/aot.factories:1]: The key in this file is wrong. Must be: `org.springframework.aot.hint.RuntimeHintsRegistrar`

File: spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java
Comment (artembilan) [inline spring-rabbit/src/main/java/org/springframework/amqp/rabbit/aot/RabbitRuntimeHintsRegistrar.java:45]: I believe we still need an individual proxy for just a `ChannelProxy.class` according to the usage logic we’ve discussed before.

**Review by artembilan — CHANGES_REQUESTED** [review]

File: spring-rabbit/src/main/resources/META-INF/spring/aot.factories
Comment (garyrussell) [reply spring-rabbit/src/main/resources/META-INF/spring/aot.factories]: Yes; I fixed that last night (but failed to re-push).

**Comment by garyrussell** [comment]
This is good to go now; smoke test works.

**Review by artembilan — APPROVED** [review]
I can merge this if there is no need in anyone else approval.

**Comment by garyrussell** [comment]
Please do; the smoke test runs fine with this.

**Pull request merged**
